console/v1/types_console_plugin.go

```diff
@@ -48,6 +48,90 @@ type ConsolePluginSpec struct {
+	//     img-src: https://img1.com/
+	//
+	//   OpenShift web console server CSP response header:
+	//     script-src: self https://script1.com/ https://script2.com/ https://script3.com/
+	//     font-src: self https://font1.com/ https://font2.com/
+	//     img-src: self https://img1.com/
+	//
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=5
+	// +listType=map
+	// +listMapKey=name
+	// +optional
+	ContentSecurityPolicy []ConsolePluginCSP `json:"contentSecurityPolicy,omitempty"`
+}
+
+// DirectiveType is an enumeration of OpenShift web console supported CSP directives.
+// LoadType is an enumeration of i18n loading types.
+// +kubebuilder:validation:Enum:=default-src;script-src;img-src;style-src;font-src
+type DirectiveType string
+
+const (
+	// default-src directive serves as a fallback for the other CSP fetch directives.
+	// For more information about the default-src directive, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/default-src
+	DefaultSrc DirectiveType = "default-src"
+	// script-src directive specifies valid sources for JavaScript.
+	// For more information about the script-src directive, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/script-src
+	ScriptSrc DirectiveType = "script-src"
+	// img-src directive specifies a valid sources of images and favicons.
+	// For more information about the img-src directive, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/img-src
+	ImgSrc DirectiveType = "img-src"
+	// style-src directive specifies valid sources for stylesheets.
+	// For more information about the style-src directive, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/style-src
+	StyleSrc DirectiveType = "style-src"
+	// font-src directive specifies valid sources for fonts loaded using @font-face.
+	// For more information about the font-src directive, see:
+	// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/font-src
+	FontSrc DirectiveType = "font-src"
+)
+
+// ConsolePluginCSP holds configuration for a specific CSP directive
+type ConsolePluginCSP struct {
+	// directive specifies which Content-Security-Policy directive to configure.
+	// Available directive types are default-src, script-src, img-src, style-src and font-src.
+	// +kubebuilder:validation:Enum:="default-src";"script-src";"img-src";"style-src";"font-src"
+	// +kubebuilder:validation:Required
+	Directive DirectiveType `json:"directive"`
+	// values defines an array of additional values to append to the console
+	// defaults for this directive.
+	// Each ConsolePlugin may define their own directives with their values.
+	// These will be set by the OpenShift web console's backend, as part of
+	// its CSP header.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems=1
+	Values []string `json:"values"`
```

Comment:
What is the format of this string? Is it always a URL? What is the maximum possible length? What characters are valid, which are not?

Reply:
@JoelSpeed I think we have to leave it open since CSP directive values are not always URLs and don't always follow a specific format.

Reply:
I agree to leave it open and add additional validation on the backend

Reply:
We probably should disallow `;` since that is a separator in the header between directives. Maybe we can do more validation on the values. Here is the syntax from the standard:

https://www.w3.org/TR/CSP2/#policy-syntax

Reply:
> directive-value = *( WSP / <VCHAR except ";" and ","> )

Looks like `;` and `,` are not allowed in values and whitespace separates values, so can be disallowed as well. `VCHAR` refers to printing characters, so I think pretty much anything else is allowed?

Reply:
@spadgett this would require following pattern
```go
// +kubebuilder:validation:Pattern=`^[\t\x20!#\$%&'\(\)\*\+\-\.\/0-9:<=>\?@A-Za-z\[\\\]\^_`\{\|\}~]*$`
```

Reply:
Does the regex here not support negative matching? ie, match not this character?

If it doesn't, you could use CEL validations quite easily, `!self.Contains(';')`, which would be readable.

> I don't see a length limit on specific values, although some HTTP servers like Apache have limits on the total size of response headers. If we pick it a limit, it would be an arbitrary value.

Even an arbitrary limit is better than none, how about starting at `1024` and seeing if you get any customers complain that it is too small?

Reply:
> Even an arbitrary limit is better than none, how about starting at 1024 and seeing if you get any customers complain that it is too small?

@JoelSpeed I agree. Works for me.